# Feather vs CSV vs Excel

## What is feather?
A Fast On-Disk Format for DataFrames for R and Python, powered by Apache Arrow. 

Feather is a fast, lightweight, and easy-to-use binary file format for storing data frames. It has a few specific design goals:

- Lightweight, minimal API: make pushing data frames in and out of memory as simple as possible

- Language agnostic: Feather files are the same whether written by Python or R code. Other languages can read and write Feather files, too.

- High read and write performance. When possible, Feather operations should be bound by local disk performance.

Feather is extremely fast. Since Feather does not currently use any compression internally

In [1]:
import numpy as np
import pandas as pd

#### Reading a temporary file having 100000 records.

In [2]:
data_csv = pd.read_csv('test.csv')

In [3]:
data_csv.shape

(100000, 31)

#### Time taken to write data into Feather.

In [4]:
%time data_csv.to_feather('test_df.feather')

Wall time: 81.1 ms


#### Time taken to write data into CSV.

In [5]:
%time data_csv.to_csv('test_df.csv')

Wall time: 3.98 s


#### Time taken to write data into Excel.

In [6]:
%time data_csv.to_excel('test_df.xlsx')

Wall time: 2min 12s


### Yes, you read that correctly: 81.1 milliseconds for writing data to feather versus approx 4 seconds for CSV and 2 minutes 12 seconds for Excel...! 😱

### Now lets checkout how much time does it take to read those files one by one.

#### Time taken to read data from Feather.

In [7]:
%time df = pd.read_feather('test_df.feather')

Wall time: 369 ms


#### Time taken to read data from CSV.

In [8]:
%time df_csv = pd.read_csv('test_df.csv')

Wall time: 803 ms


#### Time taken to read data from Excel.

In [9]:
%time df_excel = pd.read_excel('test_df.xlsx')

Wall time: 59.1 s


### Woah...!!! Just 369 milliseconds for reading feather file versus 803 milliseconds for CSV and 59 seconds for Excel.

### You can also check the size of each files which we created. Feather file takes least memory as compared to CSV and Excel.

## When not to use Feather?
Feather is not designed for long-term data storage. As there is no guarantee that the file format will be stable between versions. Instead, use Feather for quickly exchanging data between Python and R code, or for short-term storage of data frames as part of some analysis.